# Lab | Data Cleaning and Formatting

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv


# Challenge 1: Data Cleaning and Formatting

## Exercise 1: Cleaning Column Names

To ensure consistency and ease of use, standardize the column names of the dataframe. Start by taking a first look at the dataframe and identifying any column names that need to be modified. Use appropriate naming conventions and make sure that column names are descriptive and informative.

*Hint*:
- *Column names should be in lower case*
- *White spaces in column names should be replaced by `_`*
- *`st` could be replaced for `state`*

In [123]:
import pandas as pd
df=pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv')
df.head()
df.columns
df.columns = df.columns.str.lower().str.replace(" ", "_")  # Convert to lowercase and replace spaces with '_'
df = df.rename(columns={"st": "state"})  # Rename 'st' to 'state'
print(df.columns)



Index(['customer', 'state', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'policy_type', 'vehicle_class', 'total_claim_amount'],
      dtype='object')


## Exercise 2: Cleaning invalid Values

The dataset contains columns with inconsistent and incorrect values that could affect the accuracy of our analysis. Therefore, we need to clean these columns to ensure that they only contain valid data.

Note that this exercise will focus only on cleaning inconsistent values and will not involve handling null values (NaN or None).

*Hint*:
- *Gender column contains various inconsistent values such as "F", "M", "Femal", "Male", "female", which need to be standardized, for example, to "M" and "F".*
- *State abbreviations be can replaced with its full name, for example "AZ": "Arizona", "Cali": "California", "WA": "Washington"*
- *In education, "Bachelors" could be replaced by "Bachelor"*
- *In Customer Lifetime Value, delete the `%` character*
- *In vehicle class, "Sports Car", "Luxury SUV" and "Luxury Car" could be replaced by "Luxury"*

In [124]:
df_copy=df.copy()
# Standardize gender values
df_copy["gender"] = df_copy["gender"].str.lower().replace({"female": "F", "femal": "F", "f": "F", "male": "M", "m": "M"})

# Replace state abbreviations with full names
state_mapping = {"AZ": "Arizona", "Cali": "California", "WA": "Washington"}
df_copy["state"] = df_copy["state"].replace(state_mapping)

# Standardize education level
df_copy["education"] = df_copy["education"].replace({"Bachelors": "Bachelor"})

# Remove '%' from Customer Lifetime Value and convert it to numeric format
df_copy["customer_lifetime_value"] = df_copy["customer_lifetime_value"].str.replace("%", "").astype(float)

# Standardize vehicle class by grouping luxury cars
df_copy["vehicle_class"] = df_copy["vehicle_class"].replace({"Sports Car": "Luxury", "Luxury SUV": "Luxury", "Luxury Car": "Luxury"})


print(df_copy.head())


  customer       state gender             education  customer_lifetime_value  \
0  RB50392  Washington    NaN                Master                      NaN   
1  QZ44356     Arizona      F              Bachelor                697953.59   
2  AI49188      Nevada      F              Bachelor               1288743.17   
3  WW63253  California      M              Bachelor                764586.18   
4  GA49547  Washington      M  High School or Below                536307.65   

    income  monthly_premium_auto number_of_open_complaints     policy_type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   vehicle_class  total_claim_amount  
0  Four-Doo

## Exercise 3: Formatting data types

The data types of many columns in the dataset appear to be incorrect. This could impact the accuracy of our analysis. To ensure accurate analysis, we need to correct the data types of these columns. Please update the data types of the columns as appropriate.

It is important to note that this exercise does not involve handling null values (NaN or None).

*Hint*:
- *Customer lifetime value should be numeric*
- *Number of open complaints has an incorrect format. Look at the different values it takes with `unique()` and take the middle value. As an example, 1/5/00 should be 5. Number of open complaints is a string - remember you can use `split()` to deal with it and take the number you need. Finally, since it should be numeric, cast the column to be in its proper type.*

In [125]:
df_copy.dtypes

customer                      object
state                         object
gender                        object
education                     object
customer_lifetime_value      float64
income                       float64
monthly_premium_auto         float64
number_of_open_complaints     object
policy_type                   object
vehicle_class                 object
total_claim_amount           float64
dtype: object

In [126]:
df['number_of_open_complaints'].unique()

array(['1/0/00', '1/2/00', '1/1/00', '1/3/00', '1/5/00', '1/4/00', nan],
      dtype=object)

In [127]:
# Convert Customer Lifetime Value to numeric (removing '%' if present)
df_copy["customer_lifetime_value"] = df_copy["customer_lifetime_value"].astype(str).str.replace("%", "").astype(float)

# Fix Number of Open Complaints: Extract the middle number from format "X/Y/Z"
df_copy["number_of_open_complaints"] = df_copy["number_of_open_complaints"].astype(str).apply(lambda x: x.split("/")[1] if "/" in x else x).astype(float)

df_copy.dtypes

customer                      object
state                         object
gender                        object
education                     object
customer_lifetime_value      float64
income                       float64
monthly_premium_auto         float64
number_of_open_complaints    float64
policy_type                   object
vehicle_class                 object
total_claim_amount           float64
dtype: object

## Exercise 4: Dealing with Null values

Identify any columns with null or missing values. Identify how many null values each column has. You can use the `isnull()` function in pandas to find columns with null values.

Decide on a strategy for handling the null values. There are several options, including:

- Drop the rows or columns with null values
- Fill the null values with a specific value (such as the column mean or median for numerical variables, and mode for categorical variables)
- Fill the null values with the previous or next value in the column
- Fill the null values based on a more complex algorithm or model (note: we haven't covered this yet)

Implement your chosen strategy to handle the null values. You can use the `fillna()` function in pandas to fill null values or `dropna()` function to drop null values.

Verify that your strategy has successfully handled the null values. You can use the `isnull()` function again to check if there are still null values in the dataset.

Remember to document your process and explain your reasoning for choosing a particular strategy for handling null values.

After formatting data types, as a last step, convert all the numeric variables to integers.

In [128]:
# Check for missing values
missing_values = df.isnull().sum()
print(f"missing_values:\n{missing_values}")

df_copy["customer_lifetime_value"]=df_copy["customer_lifetime_value"].fillna(df_copy["customer_lifetime_value"].median())#Since customer lifetime value can vary significantly, using the median ensures that extreme values (very high or very low customer values) do not affect the dataset too much.
df_copy["income"]=df_copy["income"].fillna(df_copy["income"].median())#Income distribution is often skewed, meaning some individuals have very high incomes while others have low incomes. The median is a better representation of the central tendency than the mean.
df_copy["monthly_premium_auto"]=df_copy["monthly_premium_auto"].fillna(df_copy["monthly_premium_auto"].mean())#Monthly premiums are usually consistent and less variable than lifetime value or income. Using the mean is a reasonable choice since the values tend to follow a more normal distribution.
df_copy["number_of_open_complaints"]=df_copy["number_of_open_complaints"].fillna(df_copy["number_of_open_complaints"].mode()[0])#Since this represents the number of complaints, the most common number (mode) is a good replacement.
df_copy["total_claim_amount"]=df_copy["total_claim_amount"].fillna(df_copy["total_claim_amount"].median())#Claims can be highly variable depending on accident severity. The median avoids the influence of extreme values while still giving a good estimate

df_copy = df_copy.dropna(subset=["customer"])#it represents a unique identifier for each customer, and missing values could indicate incomplete records. If a customer entry is missing, the entire row may be unreliable.
df_copy["state"]=df_copy["state"].fillna("Unknown")#Using "Unknown" ensures we acknowledge missing data instead of making assumptions.
df_copy["gender"]=df_copy["gender"].fillna("Unknown")
df_copy["education"]=df_copy["education"].fillna(df_copy["education"].mode()[0])#Education level is categorical, and most people likely fall under a common category. Filling with the mode prevents unnecessary data loss.
df_copy["policy_type"]=df_copy["policy_type"].fillna(df_copy["policy_type"].mode()[0])
df_copy["vehicle_class"]=df_copy["vehicle_class"].fillna(df_copy["vehicle_class"].mode()[0])

print("\n missing_value after cleaning:")
print(df_copy.isnull().sum())

numeric_columns = df_copy.select_dtypes(include=["number"]).columns
df_copy[numeric_columns] = df_copy[numeric_columns].astype(int)# Convert all numeric variables to integers
print(f"\n{df_copy.dtypes}")




missing_values:
customer                     2937
state                        2937
gender                       3054
education                    2937
customer_lifetime_value      2940
income                       2937
monthly_premium_auto         2937
number_of_open_complaints    2937
policy_type                  2937
vehicle_class                2937
total_claim_amount           2937
dtype: int64

 missing_value after cleaning:
customer                     0
state                        0
gender                       0
education                    0
customer_lifetime_value      0
income                       0
monthly_premium_auto         0
number_of_open_complaints    0
policy_type                  0
vehicle_class                0
total_claim_amount           0
dtype: int64

customer                     object
state                        object
gender                       object
education                    object
customer_lifetime_value       int64
income                        

## Exercise 5: Dealing with duplicates

Use the `.duplicated()` method to identify any duplicate rows in the dataframe.

Decide on a strategy for handling the duplicates. Options include:
- Dropping all duplicate rows
- Keeping only the first occurrence of each duplicated row
- Keeping only the last occurrence of each duplicated row
- Dropping duplicates based on a subset of columns
- Dropping duplicates based on a specific column

Implement your chosen strategy using the `drop_duplicates()` function.

Verify that your strategy has successfully handled the duplicates by checking for duplicates again using `.duplicated()`.

Remember to document your process and explain your reasoning for choosing a particular strategy for handling duplicates.

Save the cleaned dataset to a new CSV file.

*Hint*: *after dropping duplicates, reset the index to ensure consistency*.

In [129]:
# Check for duplicate rows
duplicate_rows = df_copy[df_copy.duplicated()]
print(f"Number of duplicate rows: {duplicate_rows.shape[0]}")
df_copy.to_csv("cleaned_data.csv", index=False)


Number of duplicate rows: 0


# Bonus: Challenge 2: creating functions on a separate `py` file

Put all the data cleaning and formatting steps into functions, and create a main function that performs all the cleaning and formatting.

Write these functions in separate .py file(s). By putting these steps into functions, we can make the code more modular and easier to maintain.

*Hint: autoreload module is a utility module in Python that allows you to automatically reload modules in the current session when changes are made to the source code. This can be useful in situations where you are actively developing code and want to see the effects of changes you make without having to constantly restart the Python interpreter or Jupyter Notebook kernel.*

In [130]:

def load_data(filepath):
    """Loads the dataset from a CSV file."""
    return pd.read_csv(filepath)

def standardize_column_names(df):
    """Standardizes column names: lowercase and replaces spaces with underscores."""
    df.columns = df.columns.str.lower().str.replace(" ", "_")
    df.rename(columns={"st": "state"}, inplace=True)
    return df

def clean_gender(df):
    """Standardizes gender values."""
    df["gender"] = df["gender"].str.lower().replace({"female": "F", "femal": "F", "f": "F", "male": "M", "m": "M"})
    return df

def clean_state(df):
    """Replaces state abbreviations with full names."""
    state_mapping = {"AZ": "Arizona", "Cali": "California", "WA": "Washington"}
    df["state"] = df["state"].replace(state_mapping)
    return df

def clean_education(df):
    """Standardizes education levels."""
    df["education"] = df["education"].replace({"Bachelors": "Bachelor"})
    return df

def clean_customer_lifetime_value(df):
    """Removes '%' and converts Customer Lifetime Value to numeric."""
    df["customer_lifetime_value"] = df["customer_lifetime_value"].astype(str).str.replace("%", "").astype(float)
    return df

def clean_vehicle_class(df):
    """Standardizes vehicle class by grouping luxury cars."""
    df["vehicle_class"] = df["vehicle_class"].replace({"Sports Car": "Luxury", "Luxury SUV": "Luxury", "Luxury Car": "Luxury"})
    return df

def fix_open_complaints(df):
    """Extracts the middle value from 'Number of Open Complaints' format."""
    df["number_of_open_complaints"] = df["number_of_open_complaints"].astype(str).apply(lambda x: x.split("/")[1] if "/" in x else x).astype(float)
    return df

def handle_missing_values(df):
    """Handles missing values using different strategies."""
    df["customer_lifetime_value"].fillna(df["customer_lifetime_value"].median(), inplace=True)
    df["income"].fillna(df["income"].median(), inplace=True)
    df["monthly_premium_auto"].fillna(df["monthly_premium_auto"].mean(), inplace=True)
    df["number_of_open_complaints"].fillna(df["number_of_open_complaints"].mode()[0], inplace=True)
    df["total_claim_amount"].fillna(df["total_claim_amount"].median(), inplace=True)
    
    df.dropna(subset=["customer"], inplace=True)  # Drop rows where customer ID is missing
    df["state"].fillna("Unknown", inplace=True)
    df["gender"].fillna("Unknown", inplace=True)
    df["education"].fillna(df["education"].mode()[0], inplace=True)
    df["policy_type"].fillna(df["policy_type"].mode()[0], inplace=True)
    df["vehicle_class"].fillna(df["vehicle_class"].mode()[0], inplace=True)
    
    return df

def convert_numeric_columns(df):
    """Converts all numeric columns to integers."""
    numeric_columns = df.select_dtypes(include=["number"]).columns
    df[numeric_columns] = df[numeric_columns].astype(int)
    return df

def remove_duplicates(df):
    """Removes duplicate rows and resets index."""
    df = df.drop_duplicates(keep="first")
    df.reset_index(drop=True, inplace=True)
    return df

def save_cleaned_data(df, output_filepath):
    """Saves the cleaned dataset to a CSV file."""
    df.to_csv(output_filepath, index=False)

# Bonus: Challenge 3: Analyzing Clean and Formated Data

You have been tasked with analyzing the data to identify potential areas for improving customer retention and profitability. Your goal is to identify customers with a high policy claim amount and a low customer lifetime value.

In the Pandas Lab, we only looked at high policy claim amounts because we couldn't look into low customer lifetime values. If we had tried to work with that column, we wouldn't have been able to because customer lifetime value wasn't clean and in its proper format. So after cleaning and formatting the data, let's get some more interesting insights!

Instructions:

- Review the statistics again for total claim amount and customer lifetime value to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount and a low customer lifetime value. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount, and clients with a low customer lifetime value to be those in the bottom 25% of the customer lifetime value. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile and a customer lifetime value in the bottom 25th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount and low customer lifetime value data. To do so, select both columns of the dataframe simultaneously and pass it to the `.describe()` method. This will give you descriptive statistics, such as mean, median, standard deviation, minimum and maximum values for both columns at the same time, allowing you to compare and analyze their characteristics.

In [131]:
# Load the cleaned dataset
df1 = pd.read_csv("cleaned_data.csv")

# Review overall statistics for both columns
print(df1[["total_claim_amount", "customer_lifetime_value"]].describe())

       total_claim_amount  customer_lifetime_value
count         1071.000000             1.071000e+03
mean           404.528478             7.931141e+05
std            293.041850             6.426676e+05
min              0.000000             2.004350e+05
25%            201.500000             4.048530e+05
50%            354.000000             5.881740e+05
75%            532.000000             8.945205e+05
max           2893.000000             5.816655e+06


In [132]:
# Define thresholds
claim_threshold = df1["total_claim_amount"].quantile(0.75)  # 75th percentile (high claims)
clv_threshold = df1["customer_lifetime_value"].quantile(0.25)  # 25th percentile (low CLV)

# Filter customers meeting both conditions
high_claim_low_clv_customers = df1[(df1["total_claim_amount"] > claim_threshold) & (df1["customer_lifetime_value"] < clv_threshold)]


high_claim_low_clv_customers.head()

customer       state   gender education  customer_lifetime_value  income  \
17  OE15005  California  Unknown   College                   394524   28855   
23  TZ98966      Nevada  Unknown  Bachelor                   245019       0   
26  US89481  California  Unknown  Bachelor                   394637       0   
38  CM61827      Oregon  Unknown  Bachelor                   309003       0   
42  IL19217  California  Unknown  Bachelor                   393900   60021   

    monthly_premium_auto  number_of_open_complaints     policy_type  \
17                   101                          0   Personal Auto   
23                    73                          3  Corporate Auto   
26                   111                          0   Personal Auto   
38                    90                          0   Personal Auto   
42                    99                          0   Personal Auto   

    vehicle_class  total_claim_amount  
17            SUV                 647  
23  Four-Door Car                 554  
26  Four-Door Car                 799  
38   Two-Door Car                 648  
42  Four-Door Car                 882

In [133]:
# Get summary statistics for the filtered customers
print(high_claim_low_clv_customers[["total_claim_amount", "customer_lifetime_value"]].describe())

       total_claim_amount  customer_lifetime_value
count           40.000000                40.000000
mean           723.300000            341577.425000
std            149.216449             56921.466178
min            537.000000            228759.000000
25%            632.750000            304220.500000
50%            685.000000            365995.000000
75%            799.000000            387364.000000
max           1185.000000            402381.000000


Total Claim Amount

- Mean: $723 → On average, these customers claim a higher amount than the overall dataset.

- Min: $537 → The lowest claim in this group is still above the 75th percentile of the full dataset.

- Max: $1185 → Some customers claim significantly high amounts.

Customer Lifetime Value (CLV)

- Mean: $341,577 → This is well below the overall average ($793,114), confirming these customers generate less revenue.

- Min: $228,759 → Some customers have very low CLV.

- Max: $402,381 → The highest CLV in this group is still below the 25th percentile of the dataset.
